src/utils/charts.js

```diff
@@ -0,0 +1,241 @@
+  let yAxisMax;
+  // if maxActionsCount <= 100, round up yAxisMax to the nearest ten; else, to the nearest hundred
+  if (maxActionsCount <= 100) {
+    yAxisMax = Math.ceil(maxActionsCount / 10) * 10;
+  } else {
+    yAxisMax = Math.ceil(maxActionsCount / 100) * 100;
+  }
+  return yAxisMax;
+};
+
+// helper function used in getActionsByTimeOfDay
+// todo: update this function to retrieve hour of day using JS Date objects/moment
+const getActionHourOfDay = (action) => {
+  // expects action to be an object with a createdAt key
+  // createdAt should have the format "2020-12-31T23:59:59.999Z"
+  return action.createdAt.slice(11, 13);
+};
+
+// Takes array of action objects and returns an object with {key: value} pairs of {hourOfDay: #-of-actions}
+export const getActionsByTimeOfDay = (actions) => {
+  const actionsByTimeOfDay = {
+    [LATE_NIGHT]: 0,
+    [EARLY_MORNING]: 0,
+    [MORNING]: 0,
+    [AFTERNOON]: 0,
+    [EVENING]: 0,
+    [NIGHT]: 0,
+  };
+  actions.forEach((action) => {
+    const actionHourOfDay = getActionHourOfDay(action);
+    if (actionHourOfDay >= 0 && actionHourOfDay < 4) {
+      actionsByTimeOfDay[LATE_NIGHT] += 1;
+    } else if (actionHourOfDay >= 4 && actionHourOfDay < 8) {
+      actionsByTimeOfDay[EARLY_MORNING] += 1;
+    } else if (actionHourOfDay >= 8 && actionHourOfDay < 12) {
+      actionsByTimeOfDay[MORNING] += 1;
+    } else if (actionHourOfDay >= 12 && actionHourOfDay < 16) {
+      actionsByTimeOfDay[AFTERNOON] += 1;
+    } else if (actionHourOfDay >= 16 && actionHourOfDay < 20) {
+      actionsByTimeOfDay[EVENING] += 1;
+    } else if (actionHourOfDay >= 20 && actionHourOfDay < 24) {
+      actionsByTimeOfDay[NIGHT] += 1;
+    } else {
+      // eslint-disable-next-line no-console
+      console.log(`actionHourOfDay ${actionHourOfDay} is undefined`);
+    }
+  });
+  return actionsByTimeOfDay;
+};
+
+// Takes object with {key: value} pairs of {timeOfDay: #-of-actions}
+// returns a date-sorted array in Recharts.js format
+export const formatActionsByTimeOfDay = (actionsByTimeOfDayObject) => {
+  const actionsByTimeOfDayArray = Object.entries(actionsByTimeOfDayObject);
+  return actionsByTimeOfDayArray.map((entry) => {
+    return {
+      timeOfDay: entry[0],
+      count: entry[1],
+    };
+  });
+};
+
+// Takes array of action objects and returns an object with {key: value} pairs of {verb: %-of-actions}
+export const getActionsByVerb = (actions) => {
+  const totalActions = actions.length;
+  const actionsByVerb = {};
+  actions.forEach((action) => {
+    if (!actionsByVerb[action.verb]) {
+      // if verb is still not in the actionsByVerb object, add it and assign it to (1 / totalActions)
+      // we use (1 / totalActions) because in the end we want this object to be {verb: PERCENTAGE-of-total-actions}
+      actionsByVerb[action.verb] = 1 / totalActions;
+    } else {
+      actionsByVerb[action.verb] += 1 / totalActions;
+    }
+  });
+  return actionsByVerb;
```

Comment:
interesting... could do this. but downside is that we'd have to do `Object.entries()` on `actionsByVerb`, loop through that array to divide each verb's actionCount by totalActions count, and then re-convert into an object... what do you think?

Reply:
https://stackoverflow.com/questions/14810506/map-function-for-objects-instead-of-arrays
We could also iterate over the keys of the object and divide `actionsByVerb[verb] /= totalActions`. I does require a second (smaller, since we iterate over each distinct verb) loop but I think it is more correct.

Reply:
the `Object.fromEntries()` and `Object.entries()` pair work well together. fixed this.